Creado la sección de Producto para el API

MaddozS opened the pull request:
Se añadió el controlador, servicio y esquemas de este para la DB 

Se completa #2

MaddozS committed:
:heavy_plus_sign: Añadidas dependencias de mongodb

MaddozS committed:
:speech_balloon: Cambio de puerto

MaddozS committed:
:sparkles: Creado modulo de Producto

MaddozS committed:
:sparkles: Agregado modulo de Mongoose y Producto a la app

MaddozS committed:
:sparkles: Creado schema de Producto

MaddozS committed:
:sparkles: Creado DTO de Producto

MaddozS committed:
:sparkles: Creado servicio de Producto con create y findAll

MaddozS committed:
:sparkles: Creado controller de Producto

MaddozS committed:
:truck: Cambio en el import de ProductModule

alanmoraales commented on src/producto/dto/producto.dto.ts:
```diff
@@ -0,0 +1,7 @@
+export class CreateProductoDTO {
+  readonly nombre: string;
+  readonly descripcion: string;
+  readonly precio: number;
+  readonly cantidad: number;
+  readonly agregadoEl: Date;
+}
```
Veo en general que algunas palabras están en español y otras en inglés. También nombres combinados como `CreateProductoDTO`. Creo que podríamos estandarizar todo a inglés.

MaddozS replied:
Sí, pero la idea que tengo es que todo este en inglés, a excepción de los atributos y modelos, ya que estos su información en teoría tendrá que estar en español, todo lo demás que no tenga que ver con los datos como tal lo estoy dejando en inglés

alanmoraales replied:
Yo pondría los datos también en inglés, porque al final el usuario no va a verlo.

alanmoraales commented on src/producto/producto.controller.ts:
```diff
@@ -0,0 +1,28 @@
+    let producto: Producto;
+    try {
+      producto = await this.productoService.create(createProductoDTO);
+      res.status(HttpStatus.CREATED).json(producto);
```
Creo que hay un decorador para cambiar el estatus sin tener que usar `res`. Déjame reviso y te confirmo.

alanmoraales replied:
Nest tiene el decorador `HttpCode`  y una serie de [HttpExceptions](https://docs.nestjs.com/exception-filters#built-in-http-exceptions), usando esos dos podemos evitar utilizar `res`, te voy dejando más o menos como usarlo

alanmoraales commented on src/producto/schemas/producto.schema.ts:
```diff
@@ -0,0 +1,24 @@
+  nombre: string;
+
+  @Prop({ required: true })
+  descripcion: string;
```
¿Crees que sea necesaria una descripción? 🤔 En el diseño no la estoy planeando, pero podría agregarla. En todo caso podríamos dejarla como opcional.

MaddozS replied:
Puede ir la información nutrimental del producto. En general yo lo consideré importante, tal vez para algo más informal no sea necesario

alanmoraales replied:
Suena bien, lo agregaré en la UI entonces

alanmoraales commented on src/producto/schemas/producto.schema.ts:
```diff
@@ -0,0 +1,24 @@
+  precio: number;
+
+  @Prop({ required: true, isSafeInteger: true })
+  cantidad: number;
```
¿Esta propiedad hace referencia al inventario? Si es así, yo había pensado que se podía hacer un módulo a parte para el inventario y ahí poner la cantidad existente de cada producto.

MaddozS replied:
De que manera se agregaría entonces ese apartado?

alanmoraales replied:
Como lo veo, tendríamos algo como `InventoryItem` que tenga el `id` del producto, la cantidad disponible, el estado y otro atributo `minimunRecommended`, que nos va servir para el saber el estado. Si baja del mínimo, entonces el inventario de ese producto se estaría por gastar y el usuario va a saber que debe pedir más a su proveedor.

alanmoraales commented on src/producto/producto.controller.ts:
```diff
@@ -0,0 +1,28 @@
+    return productos;
+  }
+
+  @Post()
+  async create(@Res() res: Response, @Body() createProductoDTO: CreateProductoDTO) {
+    let producto: Producto;
+    try {
+      producto = await this.productoService.create(createProductoDTO);
+      res.status(HttpStatus.CREATED).json(producto);
+    } catch (err) {
+      res.status(HttpStatus.BAD_REQUEST).json(err);
+    }
```
```suggestion
  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Res() res: Response, @Body() createProductoDTO: CreateProductoDTO) {
      return await this.productoService.create(createProductoDTO);
  }
```
El `HttpCode`  se importa de `nestjs/common`

alanmoraales commented on src/producto/producto.service.ts:
```diff
@@ -0,0 +1,19 @@
+@Injectable()
+export class ProductoService {
+  constructor(@InjectModel(Producto.name) private productoModel: Model<ProductoDocument>) {}
+
+  async create(createProductoDTO: CreateProductoDTO): Promise<Producto> {
+    const createdProducto = new this.productoModel(createProductoDTO);
+    return createdProducto.save();
+  }
+
```
Aquí es donde usamos los `HttpExceptions` para mandar el `bad request`
```suggestion

  async create(createProductoDTO: CreateProductoDTO): Promise<Producto> {
    try {
      const createdProducto = new this.productoModel(createProductoDTO);
      return createdProducto.save();
    } catch (error) {
      throw new BadRequestException('El mensaje que más te guste')
    }
  }

```
`BadRequestException` también se importa desde `nestjs/common`

MaddozS replied:
Aqui en service no manejo los errores

MaddozS committed:
:truck: Estandarización a inglés del módulo Producto

MaddozS committed:
:fire: Eliminado campo "cantidad"

Este campo fue eliminado pues la cantidad del producto será manejada en otra entidad en la base de datos

MaddozS committed:
:recycle: Cambios en el manejo de errores

MaddozS committed:
:sparkles: Agregado interfaz de servicio

MaddozS committed:
:sparkles: Agregado y actualizado métodos del servicio Product

MaddozS committed:
:sparkles: Agregado y actualizado endpoints al controlador de Product

MaddozS committed:
:sparkles: Agregado codigo HTTP a las respuestas
